docker_push_latest_if_changed.py

```diff
@@ -1,5 +1,180 @@
+            'the same! Source and target tags cannot be the same.'
+        ))
+    return target
+
+
+def _has_image_changed(source, target):
+    source_key = _get_image_key(source)
+    target_key = _get_image_key(target)
+    print(f'Source key: {source_key}')
+    print(f'Target key: {target_key}')
+    return source_key != target_key
+
+
+def _get_image_key(image):
+    return _ImageKey(
+        commands_hash=_get_commands_hash(image),
+        packages_hash=_get_packages_hash(image),
+    )
+
+
+def _get_commands_hash(image):
+    image_commands = _get_image_commands(image)
+    print(f'Docker commands for {image}:\n{image_commands}')
+    return _get_sha256_hexdigest(image_commands)
+
+
+def _get_packages_hash(image):
+    packages = _run_in_image(image, ('dpkg', '-l'))
+    print(f'Packages for {image}:\n{packages}')
+    return _get_sha256_hexdigest(packages)
+
+
+def _get_sha256_hexdigest(blob):
+    return hashlib.sha256(blob).hexdigest()
+
+
+def _tag_image(source, target, *, is_dry_run):
+    print(f'Tagging image {source} as {target}')
+    tag_command = ('docker', 'tag', source, target)
+    if is_dry_run:
+        tag_command = ('#',) + tag_command
+        print(' '.join(tag_command))
+        print('Image was not actually tagged since this is a dry run')
+    else:
+        _check_output_and_print(tag_command)
+
+
+def _pull_image(image):
+    pull_command = ('docker', 'pull', image)
+    try:
+        _check_output_and_print(pull_command)
+    except subprocess.CalledProcessError as e:
+        raise ImageNotFoundError(f'The image {image} was not found') from e
+
+
+def _push_image(image, *, is_dry_run):
+    print(f'Pushing image {image} ...')
+    push_command = ('docker', 'push', image)
+    if is_dry_run:
+        push_command = ('#',) + push_command
+        print(' '.join(push_command))
+        print('Image was not actually pushed since this is a dry run')
+    else:
+        _check_output_and_print(push_command)
+
+
+def _inspect_image(image):
+    inspect_command = ('docker', 'inspect', image)
+    try:
+        docker_inspect_output = _check_output_and_print(inspect_command)
+    except subprocess.CalledProcessError as e:
+        raise ImageNotFoundError(f'The image {image} was not found') from e
+    return json.loads(docker_inspect_output)[0]
+
+
+def _get_image_commands(image):
+    history_command = (
+        'docker',
+        'history',
+        '--no-trunc',
+        '--format',
+        '{{.CreatedBy}}',
+        image,
+    )
+    return _check_output_and_print(history_command)
+
+
+def _run_in_image(image, command):
+    run_command = (
+        'docker',
+        'run',
+        '--rm',
+        '--net=none',
+        '--user=nobody',
+        image,
+    ) + command
+    return _check_output_and_print(run_command)
+
+
+def _check_output_and_print(command):
+    print(' '.join(command))
+    output = subprocess.check_output(command)
```

Comment:
this'll return `bytes`, which you then format into strings.  do you mean to `.decode('UTF-8')` or use the `encoding=` parameter of `check_output`?

Reply:
Oh yeah that actually is what I want. I didn't notice it since the prints with f strings accept bytes, but it would've looked like poop if there was actually any unicode in the output.

Did the needful with encoding.